app/Http/Controllers/API/ApiController.php

```diff
@@ -18,8 +26,68 @@
     //     new OA\ServerVariable(name: "host", default: "api.trakli.io", enum: ["api.trakli.io", "api.staging.example.com"])
     // ]
 )]
+#[OA\Components(
+    securitySchemes: [
+        new OA\SecurityScheme(
+            securityScheme: 'bearerAuth',
+            type: 'http',
+            scheme: 'bearer',
+            description: 'Bearer token authentication'
+        ),
+    ]
+)]
 class ApiController extends BaseController
 {
+    /**
+     * Check if the authenticated user can access the given resource.
+     *
+     * @param  mixed  $resource
+     * @param  string|null  $ownerKey
+     * @return void
+     *
+     * @throws \Illuminate\Http\Exceptions\HttpResponseException
+     */
+    public function userCanAccessResource($resource, $ownerKey = 'user_id')
+    {
+        if (is_null($resource)) {
+            abort(Response::HTTP_NOT_FOUND, 'Resource not found.');
+        }
+
+        if (auth()->id() !== $resource->{$ownerKey}) {
+            abort(Response::HTTP_FORBIDDEN, 'You are not authorized to access this resource.');
+        }
+    }
+
+    /**
+     * Validate the request data.
+     */
+    protected function validateRequest(Request $request, array $rules): array
+    {
+        $validator = Validator::make($request->all(), $rules);
+
+        if ($validator->fails()) {
+            $errors = $validator->errors();
+
+            if ($errors->hasAny(array_keys($rules))) {
+                return [
+                    'isValidated' => false,
+                    'message' => 'Server failed to validate request.',
+                    'code' => 422,
+                    'errors' => $errors->toArray(),
+                ];
+            }
+
+            return [
+                'isValidated' => false,
+                'message' => 'Server unable to process request.',
+                'code' => 400,
+                'errors' => $errors->toArray(),
+            ];
+        }
+
+        return ['isValidated' => true, 'data' => $validator->validated()];
+    }
+
```

Comment:
You can use [Form Requests](https://laravel.com/docs/11.x/validation#form-request-validation) for this

Reply:
Yes but, then we would have to write `FormRequests` for each endpoint! I did not want to do that.

Reply:
ok. It makes the code cleaner in my own opinion

Reply:
That's true. It usually does make things cleaner when you have large and complex requests. There is no need to have 4 different `FormRequests` (4 + files) for the CRUD operations in this file and then we would have to repeat this across all the models, in reality only a few of the requests (e.g store) are big a enough to need the `FormRequests` which I do use in other code bases but in this case I think this small function takes the bull by the horn we can use to validate all the simple requests based on the rules passed without adding any new files to the project.